Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp

```diff
@@ -0,0 +1,429 @@
+		break;
+	}
+}
+
+void msblenMaterialsExportHelper::setHeightFromOutputNode(const Material* mat, ms::StandardMaterial& stdmat, bNode* outputNode)
+{
+	auto displacementSocket = getInputSocket(outputNode, displacementIdentifier);
+	if (displacementSocket)
+	{
+		setValueFromSocket(mat,
+			displacementSocket, ms::TextureType::Default,
+			false,
+			[&](const mu::float4& colorValue) {
+				stdmat.setHeightScale(colorValue[0]);
+			},
+			[&](int textureId)
+			{
+				stdmat.setHeightMap(textureId);
+			});
+	}
+}
+
+void msblenMaterialsExportHelper::setPropertiesFromBSDF(const Material* mat, ms::StandardMaterial& stdmat, bNode* bsdfNode)
+{
+#define isSocket(NAME) STREQ(inputSocket->identifier, NAME)
+
+	// Go through all inputs on the bsdf and find colors and textures to set:
+	for (auto inputSocket : list_range((bNodeSocket*)bsdfNode->inputs.first)) {
+		if (isSocket(baseColorIdentifier) ||
+			isSocket(colorIdentifier)) {
+			setValueFromSocket(mat,
+				inputSocket, ms::TextureType::Default,
+				true,
+				[&](const mu::float4& colorValue)
+				{
+					stdmat.setColor(colorValue);
+				},
+				[&](int textureId)
+				{
+					stdmat.setColorMap(textureId);
+				});
+		}
+		else if (isSocket(roughnessIdentifier))
+		{
+			setValueFromSocket(mat,
+				inputSocket, ms::TextureType::Default,
+				false,
+				[&](const mu::float4& colorValue)
+				{
+					stdmat.setSmoothness(1 - colorValue[0]);
+				},
+				[&](int textureId) {
+					stdmat.setSmoothnessMap(textureId);
+				});
+		}
+		else if (isSocket(metallicIdentifier))
+		{
+			setValueFromSocket(mat,
+				inputSocket, ms::TextureType::Default,
+				true,
+				[&](const mu::float4& colorValue) {
+					stdmat.setMetallic(colorValue[0]);
+				},
+				[&](int textureId) {
+					stdmat.setMetallicMap(textureId);
+				});
+		}
+		else if (isSocket(normalIdentifier))
+		{
+			setValueFromSocket(mat,
+				inputSocket, ms::TextureType::NormalMap,
+				true,
+				[&](const mu::float4& colorValue) {
+					stdmat.setBumpScale(colorValue[0]);
+				},
+				[&](int textureId) {
+					stdmat.setBumpMap(textureId);
+				});
+		}
+		else if (isSocket(emissionIdentifier))
+		{
+			setValueFromSocket(mat,
+				inputSocket, ms::TextureType::Default,
+				true,
+				[&](const mu::float4& colorValue) {
+					stdmat.setEmissionColor(colorValue);
+				},
+				[&](int textureId) {
+					stdmat.setEmissionMap(textureId);
+				});
+		}
+		else if (isSocket(emissionStrengthIdentifier))
+		{
+			setValueFromSocket(mat,
+				inputSocket, ms::TextureType::Default,
+				true,
+				[&](const mu::float4& colorValue) {
+					// blender emission strength is in W/m^2
+				    // 1 W/m^2 = 683 Lumen/m^2 (Lux)
+					stdmat.setEmissionStrength(colorValue[0] * 683);
+				},
+				nullptr);
+		}
+	}
+}
+
+void msblenMaterialsExportHelper::exportMaterialFromNodeTree(const Material* mat, ms::StandardMaterial& stdmat)
+{
+	bNode* bsdfNode;
+	bNode* outputNode;
+
+	if (!getBSDFAndOutput(mat, bsdfNode, outputNode)) {
+		return;
```

Comment:
Is there any indication about what's wrong to the user if the material sync doesn't work ?

Reply:
This can only return false if there is no material output node or no bsdf connected to the material output. 
Every material should have these or it would just be black.

Reply:
And even in general, if the material sync fails, do we let the user know if there are nodes or things we can't handle?

Reply:
Would we want to show that kind of message in blender or Unity?

Reply:
I suppose we wouldn't really have knowledge of why the material sync failed in Unity unless we sent over a specific failure message to display there...

Reply:
Yeah that's what I'm wondering. We can implement sending text messages to Unity or just show the error in blender. In Unity we could log it as an error, in blender there's no unobtrusive way but still obvious way to show it.